### PR TITLE
[Serve] Fixing `DeploymentStateManager` qualifying replicas as running prematurely

### LINKED ISSUE #43034: [Serve] When downscaling Serve prefers RUNNING nodes over STARTING ones
### What happened + What you expected to happen

In our Serve application, replicas `__init__` sequence could take anywhere from 1-10 min to boot up.

What we've noticed is that when downscaling Serve tends to sometime prefer killing `RUNNING` replicas over `STARTING` ones, even though i believe it's actually supposed to be preferring `STARTING` and `STOPPING` ones.

### Versions / Dependencies

2.9.1

### Reproduction script

Run Serve app withe deployment with multiple replicas, where each replica has slow `__init__`
Trigger upscaling, then immediately trigger downscaling
Serve seems to occasionally kill perfectly fine running replica causing a downtime even though there was starting replica that could have been axed instead

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by alexeykudinkin**
Context of the issue is captured in: https://github.com/ray-project/ray/issues/43034

## Why are these changes needed?

Currently, `DeploymentStateManager` is qualifying replicas as running even thought they aren't _actually_ running yet. 

This entails issues during downscaling: 

1. `DefaultDeploymentScheduler` is supposed to prefer `STARTING`/`STOPPING`/etc replicas during downscaling for stopping
2. With replicas marked as replicas prematurely, `DeploymentScheduler` in some cases would pick *actually* running replicas for stopping even though there are still ones that aren't actually running yet

## Related issue number

Addresses https://github.com/ray-project/ray/issues/43034

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 1]
```diff
@@ -265,51 +280,6 @@ def check_health(self):
         return self.healthy
 
 
-class MockDeploymentScheduler:
```
Comment (alexeykudinkin) [inline python/ray/serve/tests/unit/test_deployment_state.py:268]: Killing this to make sure we actually run production one in our tests

Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:268]: hallelujah

File: python/ray/serve/_private/deployment_state.py [hunk 2]
```diff
@@ -1012,9 +1012,11 @@ def stop(self, graceful: bool = True) -> None:
 
         Should handle the case where the replica is already stopped.
         """
+        state = self._actor_details.state
         logger.info(
-            f"Stopping replica {self.replica_tag} for deployment "
-            f"'{self.deployment_name}' in application '{self.app_name}'.",
+            f"Stopping replica {self.replica_tag} (currently {state}) "
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:1017]: does this render in a human-readable format?

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 3]
```diff
@@ -367,6 +337,8 @@ def mock_deployment_state() -> Tuple[DeploymentState, Mock, Mock]:
         "ray.serve._private.deployment_state.ActorReplicaWrapper",
         new=MockReplicaActorWrapper,
     ), patch("time.time", new=timer.time), patch(
+        "ray.serve._private.deployment_scheduler.get_head_node_id",
+    ) as mock_get_head_node_id, patch(
```
Comment (edoakes) [inline python/ray/serve/tests/unit/test_deployment_state.py:341]: No need for this mock, let's just pass the head node ID into the replica scheduler constructor (the function is only called once in its constructor)

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 4]
```diff
@@ -2372,6 +2348,146 @@ def test_autoscale(mock_deployment_state_manager_full, target_capacity_direction
...
+    assert deployment_state.curr_status_info.status == DeploymentStatus.UPSCALING
+    assert (
+        deployment_state.curr_status_info.status_trigger
+        == DeploymentStatusTrigger.AUTOSCALING
+    )
+
+    # Set replicas as PENDING_INITIALIZATION: actors have been successfully allocated,
+    # but replicas are still pending successful initialization
+    for replica in deployment_state._replicas.get():
+        replica._actor.set_pending_init()
```
Comment (edoakes) [inline python/ray/serve/tests/unit/test_deployment_state.py:2430]: can we please also test the case for `PENDING_ALLOCATION`? you could just parametrize the test for each case

(in the future, we should also verify that `PENDING_ALLOCATION` is preferred over `PENDING_INITIALIZATION`)

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 5]
```diff
@@ -3018,7 +3134,9 @@ def mock_deployment_state_manager_full(
         "ray.serve._private.long_poll.LongPollHost"
     ) as mock_long_poll, patch(
         "ray.get_runtime_context"
-    ):
+    ), patch(
+        "ray.serve._private.deployment_scheduler.get_head_node_id",
+    ) as mock_get_head_node_id:
```
Comment (edoakes) [inline python/ray/serve/tests/unit/test_deployment_state.py:3139]: ditto from above, let's just pass it in

**Review by edoakes — COMMENTED** [review]
Looks great thanks for the fix, just small comments

File: python/ray/serve/_private/deployment_state.py (hunk 2 above)
Comment (alexeykudinkin) [reply python/ray/serve/_private/deployment_state.py:1017]: It renders beautifully:

```
2024-02-08 11:41:44,461	INFO deployment_state.py:1916 -- Removing 3 replicas from deployment 'deployment_with_slow_to_start_replicas' in application 'test_app'.
2024-02-08 11:41:44,461	INFO deployment_state.py:1015 -- Stopping replica test_app#deployment_with_slow_to_start_replicas#l7tbrjfb (currently STARTING) for deployment 'deployment_with_slow_to_start_replicas' in application 'test_app'.
2024-02-08 11:41:44,461	INFO deployment_state.py:1015 -- Stopping replica test_app#deployment_with_slow_to_start_replicas#6wjfpi2f (currently RUNNING) for deployment 'deployment_with_slow_to_start_replicas' in application 'test_app'.
2024-02-08 11:41:44,461	INFO deployment_state.py:1015 -- Stopping replica test_app#deployment_with_slow_to_start_replicas#x76wroek (currently RUNNING) for deployment 'deployment_with_slow_to_start_replicas' in application 'test_app'.
```

Comment (alexeykudinkin) [reply python/ray/serve/_private/deployment_state.py:1017]: This log is actually from the test run on master, which is obviously failing ^

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 6]
```diff
@@ -203,14 +209,23 @@ def set_node_id(self, node_id: str):
 
     def start(self, deployment_info: DeploymentInfo):
         self.started = True
+
+        def _on_scheduled_stub(*args, **kwargs):
+            print(
+                f"ReplicaSchedulingRequest.on_scheduled was invoked with:\n"
+                f"args={args}\n"
+                f"kwargs={kwargs}"
```
[the remark this answers is not in the document]
Comment (alexeykudinkin) [reply python/ray/serve/tests/unit/test_deployment_state.py:217]: That unfortunately exceeds the line length

**Comment by alexeykudinkin** [comment]
@edoakes addressed your comments, PTAL

**Review by edoakes — COMMENTED** [review]
LGTM aside from change to `default_impl.py`, let's sync on it offline

File: python/ray/serve/_private/default_impl.py [hunk 7]
```diff
@@ -7,6 +9,13 @@
     DefaultDeploymentScheduler,
     DeploymentScheduler,
 )
+from ray.serve._private.utils import get_head_node_id
+
+# NOTE: Please read carefully before changing!
+#
+# These methods are common extension points, therefore these should be
+# changed as a Developer API, ie methods should not be renamed, have their
+# API modified w/o substantial enough justification
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/default_impl.py:18]:
```suggestion
# treated as a Developer API, ie methods should not be renamed, have their
# API modified w/o substantial enough justification
```

**Review by edoakes — APPROVED** [review]

**Review by jjyao — APPROVED** [review]

**Comment by edoakes** [comment]
Test faillure @alexeykudinkin

**Pull request merged**